// @aws-sdk/client-acm/dist-cjs/index.js
'use strict';

var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');
var middlewareSerde = require('@smithy/middleware-serde');
var core$1 = require('@aws-sdk/core');
var utilWaiter = require('@smithy/util-waiter');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "acm",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class ACMClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultACMHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class ACMServiceException extends smithyClient.ServiceException {
    constructor(options) {
        super(options);
        Object.setPrototypeOf(this, ACMServiceException.prototype);
    }
}

class AccessDeniedException extends ACMServiceException {
    name = "AccessDeniedException";
    $fault = "client";
    Message;
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
        this.Message = opts.Message;
    }
}
class InvalidArnException extends ACMServiceException {
    name = "InvalidArnException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidArnException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidArnException.prototype);
    }
}
class InvalidParameterException extends ACMServiceException {
    name = "InvalidParameterException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
    }
}
class InvalidTagException extends ACMServiceException {
    name = "InvalidTagException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidTagException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidTagException.prototype);
    }
}
class ResourceNotFoundException extends ACMServiceException {
    name = "ResourceNotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
    }
}
class TagPolicyException extends ACMServiceException {
    name = "TagPolicyException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "TagPolicyException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TagPolicyException.prototype);
    }
}
class ThrottlingException extends ACMServiceException {
    name = "ThrottlingException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
    }
}
class TooManyTagsException extends ACMServiceException {
    name = "TooManyTagsException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "TooManyTagsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, TooManyTagsException.prototype);
    }
}
const RecordType = {
    CNAME: "CNAME",
};
const ValidationMethod = {
    DNS: "DNS",
    EMAIL: "EMAIL",
    HTTP: "HTTP",
};
const DomainStatus = {
    FAILED: "FAILED",
    PENDING_VALIDATION: "PENDING_VALIDATION",
    SUCCESS: "SUCCESS",
};
const ExtendedKeyUsageName = {
    ANY: "ANY",
    CODE_SIGNING: "CODE_SIGNING",
    CUSTOM: "CUSTOM",
    EMAIL_PROTECTION: "EMAIL_PROTECTION",
    IPSEC_END_SYSTEM: "IPSEC_END_SYSTEM",
    IPSEC_TUNNEL: "IPSEC_TUNNEL",
    IPSEC_USER: "IPSEC_USER",
    NONE: "NONE",
    OCSP_SIGNING: "OCSP_SIGNING",
    TIME_STAMPING: "TIME_STAMPING",
    TLS_WEB_CLIENT_AUTHENTICATION: "TLS_WEB_CLIENT_AUTHENTICATION",
    TLS_WEB_SERVER_AUTHENTICATION: "TLS_WEB_SERVER_AUTHENTICATION",
};
const FailureReason = {
    ADDITIONAL_VERIFICATION_REQUIRED: "ADDITIONAL_VERIFICATION_REQUIRED",
    CAA_ERROR: "CAA_ERROR",
    DOMAIN_NOT_ALLOWED: "DOMAIN_NOT_ALLOWED",
    DOMAIN_VALIDATION_DENIED: "DOMAIN_VALIDATION_DENIED",
    INVALID_PUBLIC_DOMAIN: "INVALID_PUBLIC_DOMAIN",
    NO_AVAILABLE_CONTACTS: "NO_AVAILABLE_CONTACTS",
    OTHER: "OTHER",
    PCA_ACCESS_DENIED: "PCA_ACCESS_DENIED",
    PCA_INVALID_ARGS: "PCA_INVALID_ARGS",
    PCA_INVALID_ARN: "PCA_INVALID_ARN",
    PCA_INVALID_DURATION: "PCA_INVALID_DURATION",
    PCA_INVALID_STATE: "PCA_INVALID_STATE",
    PCA_LIMIT_EXCEEDED: "PCA_LIMIT_EXCEEDED",
    PCA_NAME_CONSTRAINTS_VALIDATION: "PCA_NAME_CONSTRAINTS_VALIDATION",
    PCA_REQUEST_FAILED: "PCA_REQUEST_FAILED",
    PCA_RESOURCE_NOT_FOUND: "PCA_RESOURCE_NOT_FOUND",
    SLR_NOT_FOUND: "SLR_NOT_FOUND",
};
const KeyAlgorithm = {
    EC_prime256v1: "EC_prime256v1",
    EC_secp384r1: "EC_secp384r1",
    EC_secp521r1: "EC_secp521r1",
    RSA_1024: "RSA_1024",
    RSA_2048: "RSA_2048",
    RSA_3072: "RSA_3072",
    RSA_4096: "RSA_4096",
};
const KeyUsageName = {
    ANY: "ANY",
    CERTIFICATE_SIGNING: "CERTIFICATE_SIGNING",
    CRL_SIGNING: "CRL_SIGNING",
    CUSTOM: "CUSTOM",
    DATA_ENCIPHERMENT: "DATA_ENCIPHERMENT",
    DECIPHER_ONLY: "DECIPHER_ONLY",
    DIGITAL_SIGNATURE: "DIGITAL_SIGNATURE",
    ENCHIPER_ONLY: "ENCIPHER_ONLY",
    KEY_AGREEMENT: "KEY_AGREEMENT",
    KEY_ENCIPHERMENT: "KEY_ENCIPHERMENT",
    NON_REPUDATION: "NON_REPUDIATION",
};
const CertificateManagedBy = {
    CLOUDFRONT: "CLOUDFRONT",
};
const CertificateTransparencyLoggingPreference = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
const CertificateExport = {
    DISABLED: "DISABLED",
    ENABLED: "ENABLED",
};
const RenewalEligibility = {
    ELIGIBLE: "ELIGIBLE",
    INELIGIBLE: "INELIGIBLE",
};
const RenewalStatus = {
    FAILED: "FAILED",
    PENDING_AUTO_RENEWAL: "PENDING_AUTO_RENEWAL",
    PENDING_VALIDATION: "PENDING_VALIDATION",
    SUCCESS: "SUCCESS",
};
const RevocationReason = {
    AFFILIATION_CHANGED: "AFFILIATION_CHANGED",
    A_A_COMPROMISE: "A_A_COMPROMISE",
    CA_COMPROMISE: "CA_COMPROMISE",
    CERTIFICATE_HOLD: "CERTIFICATE_HOLD",
    CESSATION_OF_OPERATION: "CESSATION_OF_OPERATION",
    KEY_COMPROMISE: "KEY_COMPROMISE",
    PRIVILEGE_WITHDRAWN: "PRIVILEGE_WITHDRAWN",
    REMOVE_FROM_CRL: "REMOVE_FROM_CRL",
    SUPERCEDED: "SUPERCEDED",
    SUPERSEDED: "SUPERSEDED",
    UNSPECIFIED: "UNSPECIFIED",
};
const CertificateStatus = {
    EXPIRED: "EXPIRED",
    FAILED: "FAILED",
    INACTIVE: "INACTIVE",
    ISSUED: "ISSUED",
    PENDING_VALIDATION: "PENDING_VALIDATION",
    REVOKED: "REVOKED",
    VALIDATION_TIMED_OUT: "VALIDATION_TIMED_OUT",
};
const CertificateType = {
    AMAZON_ISSUED: "AMAZON_ISSUED",
    IMPORTED: "IMPORTED",
    PRIVATE: "PRIVATE",
};
class ConflictException extends ACMServiceException {
    name = "ConflictException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ConflictException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ConflictException.prototype);
    }
}
class ResourceInUseException extends ACMServiceException {
    name = "ResourceInUseException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourceInUseException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceInUseException.prototype);
    }
}
class RequestInProgressException extends ACMServiceException {
    name = "RequestInProgressException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "RequestInProgressException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, RequestInProgressException.prototype);
    }
}
class LimitExceededException extends ACMServiceException {
    name = "LimitExceededException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "LimitExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, LimitExceededException.prototype);
    }
}
class InvalidArgsException extends ACMServiceException {
    name = "InvalidArgsException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidArgsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidArgsException.prototype);
    }
}
const SortBy = {
    CREATED_AT: "CREATED_AT",
};
const SortOrder = {
    ASCENDING: "ASCENDING",
    DESCENDING: "DESCENDING",
};
class ValidationException extends ACMServiceException {
    name = "ValidationException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ValidationException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ValidationException.prototype);
    }
}
class InvalidDomainValidationOptionsException extends ACMServiceException {
    name = "InvalidDomainValidationOptionsException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidDomainValidationOptionsException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidDomainValidationOptionsException.prototype);
    }
}
class InvalidStateException extends ACMServiceException {
    name = "InvalidStateException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "InvalidStateException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidStateException.prototype);
    }
}
const ExportCertificateRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.Passphrase && { Passphrase: smithyClient.SENSITIVE_STRING }),
});
const ExportCertificateResponseFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.PrivateKey && { PrivateKey: smithyClient.SENSITIVE_STRING }),
});
const ImportCertificateRequestFilterSensitiveLog = (obj) => ({
    ...obj,
    ...(obj.PrivateKey && { PrivateKey: smithyClient.SENSITIVE_STRING }),
});

const se_AddTagsToCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("AddTagsToCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DeleteCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("DeleteCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_DescribeCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("DescribeCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ExportCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("ExportCertificate");
    let body;
    body = JSON.stringify(se_ExportCertificateRequest(input, context));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_GetAccountConfigurationCommand = async (input, context) => {
    const headers = sharedHeaders("GetAccountConfiguration");
    const body = "{}";
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_GetCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("GetCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ImportCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("ImportCertificate");
    let body;
    body = JSON.stringify(se_ImportCertificateRequest(input, context));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListCertificatesCommand = async (input, context) => {
    const headers = sharedHeaders("ListCertificates");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ListTagsForCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("ListTagsForCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_PutAccountConfigurationCommand = async (input, context) => {
    const headers = sharedHeaders("PutAccountConfiguration");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_RemoveTagsFromCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("RemoveTagsFromCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_RenewCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("RenewCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_RequestCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("RequestCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_ResendValidationEmailCommand = async (input, context) => {
    const headers = sharedHeaders("ResendValidationEmail");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_RevokeCertificateCommand = async (input, context) => {
    const headers = sharedHeaders("RevokeCertificate");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const se_UpdateCertificateOptionsCommand = async (input, context) => {
    const headers = sharedHeaders("UpdateCertificateOptions");
    let body;
    body = JSON.stringify(smithyClient._json(input));
    return buildHttpRpcRequest(context, headers, "/", undefined, body);
};
const de_AddTagsToCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DeleteCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_DescribeCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_DescribeCertificateResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ExportCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_GetAccountConfigurationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_GetCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ImportCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListCertificatesCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = de_ListCertificatesResponse(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ListTagsForCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_PutAccountConfigurationCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_RemoveTagsFromCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_RenewCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_RequestCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_ResendValidationEmailCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_RevokeCertificateCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    const data = await core$1.parseJsonBody(output.body, context);
    let contents = {};
    contents = smithyClient._json(data);
    const response = {
        $metadata: deserializeMetadata(output),
        ...contents,
    };
    return response;
};
const de_UpdateCertificateOptionsCommand = async (output, context) => {
    if (output.statusCode >= 300) {
        return de_CommandError(output, context);
    }
    await smithyClient.collectBody(output.body, context);
    const response = {
        $metadata: deserializeMetadata(output),
    };
    return response;
};
const de_CommandError = async (output, context) => {
    const parsedOutput = {
        ...output,
        body: await core$1.parseJsonErrorBody(output.body, context),
    };
    const errorCode = core$1.loadRestJsonErrorCode(output, parsedOutput.body);
    switch (errorCode) {
        case "InvalidArnException":
        case "com.amazonaws.acm#InvalidArnException":
            throw await de_InvalidArnExceptionRes(parsedOutput);
        case "InvalidParameterException":
        case "com.amazonaws.acm#InvalidParameterException":
            throw await de_InvalidParameterExceptionRes(parsedOutput);
        case "InvalidTagException":
        case "com.amazonaws.acm#InvalidTagException":
            throw await de_InvalidTagExceptionRes(parsedOutput);
        case "ResourceNotFoundException":
        case "com.amazonaws.acm#ResourceNotFoundException":
            throw await de_ResourceNotFoundExceptionRes(parsedOutput);
        case "TagPolicyException":
        case "com.amazonaws.acm#TagPolicyException":
            throw await de_TagPolicyExceptionRes(parsedOutput);
        case "ThrottlingException":
        case "com.amazonaws.acm#ThrottlingException":
            throw await de_ThrottlingExceptionRes(parsedOutput);
        case "TooManyTagsException":
        case "com.amazonaws.acm#TooManyTagsException":
            throw await de_TooManyTagsExceptionRes(parsedOutput);
        case "AccessDeniedException":
        case "com.amazonaws.acm#AccessDeniedException":
            throw await de_AccessDeniedExceptionRes(parsedOutput);
        case "ConflictException":
        case "com.amazonaws.acm#ConflictException":
            throw await de_ConflictExceptionRes(parsedOutput);
        case "ResourceInUseException":
        case "com.amazonaws.acm#ResourceInUseException":
            throw await de_ResourceInUseExceptionRes(parsedOutput);
        case "RequestInProgressException":
        case "com.amazonaws.acm#RequestInProgressException":
            throw await de_RequestInProgressExceptionRes(parsedOutput);
        case "LimitExceededException":
        case "com.amazonaws.acm#LimitExceededException":
            throw await de_LimitExceededExceptionRes(parsedOutput);
        case "InvalidArgsException":
        case "com.amazonaws.acm#InvalidArgsException":
            throw await de_InvalidArgsExceptionRes(parsedOutput);
        case "ValidationException":
        case "com.amazonaws.acm#ValidationException":
            throw await de_ValidationExceptionRes(parsedOutput);
        case "InvalidDomainValidationOptionsException":
        case "com.amazonaws.acm#InvalidDomainValidationOptionsException":
            throw await de_InvalidDomainValidationOptionsExceptionRes(parsedOutput);
        case "InvalidStateException":
        case "com.amazonaws.acm#InvalidStateException":
            throw await de_InvalidStateExceptionRes(parsedOutput);
        default:
            const parsedBody = parsedOutput.body;
            return throwDefaultError({
                output,
                parsedBody,
                errorCode,
            });
    }
};
const de_AccessDeniedExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new AccessDeniedException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ConflictExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ConflictException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidArgsExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidArgsException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidArnExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidArnException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidDomainValidationOptionsExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidDomainValidationOptionsException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidParameterExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidParameterException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidStateExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidStateException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_InvalidTagExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new InvalidTagException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_LimitExceededExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new LimitExceededException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_RequestInProgressExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new RequestInProgressException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ResourceInUseExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ResourceInUseException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ResourceNotFoundExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ResourceNotFoundException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_TagPolicyExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new TagPolicyException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ThrottlingExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ThrottlingException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_TooManyTagsExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new TooManyTagsException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const de_ValidationExceptionRes = async (parsedOutput, context) => {
    const body = parsedOutput.body;
    const deserialized = smithyClient._json(body);
    const exception = new ValidationException({
        $metadata: deserializeMetadata(parsedOutput),
        ...deserialized,
    });
    return smithyClient.decorateServiceException(exception, body);
};
const se_ExportCertificateRequest = (input, context) => {
    return smithyClient.take(input, {
        CertificateArn: [],
        Passphrase: context.base64Encoder,
    });
};
const se_ImportCertificateRequest = (input, context) => {
    return smithyClient.take(input, {
        Certificate: context.base64Encoder,
        CertificateArn: [],
        CertificateChain: context.base64Encoder,
        PrivateKey: context.base64Encoder,
        Tags: smithyClient._json,
    });
};
const de_CertificateDetail = (output, context) => {
    return smithyClient.take(output, {
        CertificateArn: smithyClient.expectString,
        CertificateAuthorityArn: smithyClient.expectString,
        CreatedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        DomainName: smithyClient.expectString,
        DomainValidationOptions: smithyClient._json,
        ExtendedKeyUsages: smithyClient._json,
        FailureReason: smithyClient.expectString,
        ImportedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        InUseBy: smithyClient._json,
        IssuedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Issuer: smithyClient.expectString,
        KeyAlgorithm: smithyClient.expectString,
        KeyUsages: smithyClient._json,
        ManagedBy: smithyClient.expectString,
        NotAfter: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        NotBefore: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Options: smithyClient._json,
        RenewalEligibility: smithyClient.expectString,
        RenewalSummary: (_) => de_RenewalSummary(_),
        RevocationReason: smithyClient.expectString,
        RevokedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Serial: smithyClient.expectString,
        SignatureAlgorithm: smithyClient.expectString,
        Status: smithyClient.expectString,
        Subject: smithyClient.expectString,
        SubjectAlternativeNames: smithyClient._json,
        Type: smithyClient.expectString,
    });
};
const de_CertificateSummary = (output, context) => {
    return smithyClient.take(output, {
        CertificateArn: smithyClient.expectString,
        CreatedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        DomainName: smithyClient.expectString,
        ExportOption: smithyClient.expectString,
        Exported: smithyClient.expectBoolean,
        ExtendedKeyUsages: smithyClient._json,
        HasAdditionalSubjectAlternativeNames: smithyClient.expectBoolean,
        ImportedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        InUse: smithyClient.expectBoolean,
        IssuedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        KeyAlgorithm: smithyClient.expectString,
        KeyUsages: smithyClient._json,
        ManagedBy: smithyClient.expectString,
        NotAfter: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        NotBefore: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        RenewalEligibility: smithyClient.expectString,
        RevokedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
        Status: smithyClient.expectString,
        SubjectAlternativeNameSummaries: smithyClient._json,
        Type: smithyClient.expectString,
    });
};
const de_CertificateSummaryList = (output, context) => {
    const retVal = (output || [])
        .filter((e) => e != null)
        .map((entry) => {
        return de_CertificateSummary(entry);
    });
    return retVal;
};
const de_DescribeCertificateResponse = (output, context) => {
    return smithyClient.take(output, {
        Certificate: (_) => de_CertificateDetail(_),
    });
};
const de_ListCertificatesResponse = (output, context) => {
    return smithyClient.take(output, {
        CertificateSummaryList: (_) => de_CertificateSummaryList(_),
        NextToken: smithyClient.expectString,
    });
};
const de_RenewalSummary = (output, context) => {
    return smithyClient.take(output, {
        DomainValidationOptions: smithyClient._json,
        RenewalStatus: smithyClient.expectString,
        RenewalStatusReason: smithyClient.expectString,
        UpdatedAt: (_) => smithyClient.expectNonNull(smithyClient.parseEpochTimestamp(smithyClient.expectNumber(_))),
    });
};
const deserializeMetadata = (output) => ({
    httpStatusCode: output.statusCode,
    requestId: output.headers["x-amzn-requestid"] ?? output.headers["x-amzn-request-id"] ?? output.headers["x-amz-request-id"],
    extendedRequestId: output.headers["x-amz-id-2"],
    cfId: output.headers["x-amz-cf-id"],
});
const throwDefaultError = smithyClient.withBaseException(ACMServiceException);
const buildHttpRpcRequest = async (context, headers, path, resolvedHostname, body) => {
    const { hostname, protocol = "https", port, path: basePath } = await context.endpoint();
    const contents = {
        protocol,
        hostname,
        port,
        method: "POST",
        path: basePath.endsWith("/") ? basePath.slice(0, -1) + path : basePath + path,
        headers,
    };
    if (body !== undefined) {
        contents.body = body;
    }
    return new protocolHttp.HttpRequest(contents);
};
function sharedHeaders(operation) {
    return {
        "content-type": "application/x-amz-json-1.1",
        "x-amz-target": `CertificateManager.${operation}`,
    };
}

class AddTagsToCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "AddTagsToCertificate", {})
    .n("ACMClient", "AddTagsToCertificateCommand")
    .f(void 0, void 0)
    .ser(se_AddTagsToCertificateCommand)
    .de(de_AddTagsToCertificateCommand)
    .build() {
}

class DeleteCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "DeleteCertificate", {})
    .n("ACMClient", "DeleteCertificateCommand")
    .f(void 0, void 0)
    .ser(se_DeleteCertificateCommand)
    .de(de_DeleteCertificateCommand)
    .build() {
}

class DescribeCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "DescribeCertificate", {})
    .n("ACMClient", "DescribeCertificateCommand")
    .f(void 0, void 0)
    .ser(se_DescribeCertificateCommand)
    .de(de_DescribeCertificateCommand)
    .build() {
}

class ExportCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "ExportCertificate", {})
    .n("ACMClient", "ExportCertificateCommand")
    .f(ExportCertificateRequestFilterSensitiveLog, ExportCertificateResponseFilterSensitiveLog)
    .ser(se_ExportCertificateCommand)
    .de(de_ExportCertificateCommand)
    .build() {
}

class GetAccountConfigurationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "GetAccountConfiguration", {})
    .n("ACMClient", "GetAccountConfigurationCommand")
    .f(void 0, void 0)
    .ser(se_GetAccountConfigurationCommand)
    .de(de_GetAccountConfigurationCommand)
    .build() {
}

class GetCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "GetCertificate", {})
    .n("ACMClient", "GetCertificateCommand")
    .f(void 0, void 0)
    .ser(se_GetCertificateCommand)
    .de(de_GetCertificateCommand)
    .build() {
}

class ImportCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "ImportCertificate", {})
    .n("ACMClient", "ImportCertificateCommand")
    .f(ImportCertificateRequestFilterSensitiveLog, void 0)
    .ser(se_ImportCertificateCommand)
    .de(de_ImportCertificateCommand)
    .build() {
}

class ListCertificatesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "ListCertificates", {})
    .n("ACMClient", "ListCertificatesCommand")
    .f(void 0, void 0)
    .ser(se_ListCertificatesCommand)
    .de(de_ListCertificatesCommand)
    .build() {
}

class ListTagsForCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "ListTagsForCertificate", {})
    .n("ACMClient", "ListTagsForCertificateCommand")
    .f(void 0, void 0)
    .ser(se_ListTagsForCertificateCommand)
    .de(de_ListTagsForCertificateCommand)
    .build() {
}

class PutAccountConfigurationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "PutAccountConfiguration", {})
    .n("ACMClient", "PutAccountConfigurationCommand")
    .f(void 0, void 0)
    .ser(se_PutAccountConfigurationCommand)
    .de(de_PutAccountConfigurationCommand)
    .build() {
}

class RemoveTagsFromCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "RemoveTagsFromCertificate", {})
    .n("ACMClient", "RemoveTagsFromCertificateCommand")
    .f(void 0, void 0)
    .ser(se_RemoveTagsFromCertificateCommand)
    .de(de_RemoveTagsFromCertificateCommand)
    .build() {
}

class RenewCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "RenewCertificate", {})
    .n("ACMClient", "RenewCertificateCommand")
    .f(void 0, void 0)
    .ser(se_RenewCertificateCommand)
    .de(de_RenewCertificateCommand)
    .build() {
}

class RequestCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "RequestCertificate", {})
    .n("ACMClient", "RequestCertificateCommand")
    .f(void 0, void 0)
    .ser(se_RequestCertificateCommand)
    .de(de_RequestCertificateCommand)
    .build() {
}

class ResendValidationEmailCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "ResendValidationEmail", {})
    .n("ACMClient", "ResendValidationEmailCommand")
    .f(void 0, void 0)
    .ser(se_ResendValidationEmailCommand)
    .de(de_ResendValidationEmailCommand)
    .build() {
}

class RevokeCertificateCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "RevokeCertificate", {})
    .n("ACMClient", "RevokeCertificateCommand")
    .f(void 0, void 0)
    .ser(se_RevokeCertificateCommand)
    .de(de_RevokeCertificateCommand)
    .build() {
}

class UpdateCertificateOptionsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [
        middlewareSerde.getSerdePlugin(config, this.serialize, this.deserialize),
        middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions()),
    ];
})
    .s("CertificateManager", "UpdateCertificateOptions", {})
    .n("ACMClient", "UpdateCertificateOptionsCommand")
    .f(void 0, void 0)
    .ser(se_UpdateCertificateOptionsCommand)
    .de(de_UpdateCertificateOptionsCommand)
    .build() {
}

const commands = {
    AddTagsToCertificateCommand,
    DeleteCertificateCommand,
    DescribeCertificateCommand,
    ExportCertificateCommand,
    GetAccountConfigurationCommand,
    GetCertificateCommand,
    ImportCertificateCommand,
    ListCertificatesCommand,
    ListTagsForCertificateCommand,
    PutAccountConfigurationCommand,
    RemoveTagsFromCertificateCommand,
    RenewCertificateCommand,
    RequestCertificateCommand,
    ResendValidationEmailCommand,
    RevokeCertificateCommand,
    UpdateCertificateOptionsCommand,
};
class ACM extends ACMClient {
}
smithyClient.createAggregatedClient(commands, ACM);

const paginateListCertificates = core.createPaginator(ACMClient, ListCertificatesCommand, "NextToken", "NextToken", "MaxItems");

const checkState = async (client, input) => {
    let reason;
    try {
        const result = await client.send(new DescribeCertificateCommand(input));
        reason = result;
        try {
            const returnComparator = () => {
                const flat_1 = [].concat(...result.Certificate.DomainValidationOptions);
                const projection_3 = flat_1.map((element_2) => {
                    return element_2.ValidationStatus;
                });
                return projection_3;
            };
            let allStringEq_5 = returnComparator().length > 0;
            for (const element_4 of returnComparator()) {
                allStringEq_5 = allStringEq_5 && element_4 == "SUCCESS";
            }
            if (allStringEq_5) {
                return { state: utilWaiter.WaiterState.SUCCESS, reason };
            }
        }
        catch (e) { }
        try {
            const returnComparator = () => {
                const flat_1 = [].concat(...result.Certificate.DomainValidationOptions);
                const projection_3 = flat_1.map((element_2) => {
                    return element_2.ValidationStatus;
                });
                return projection_3;
            };
            for (const anyStringEq_4 of returnComparator()) {
                if (anyStringEq_4 == "PENDING_VALIDATION") {
                    return { state: utilWaiter.WaiterState.RETRY, reason };
                }
            }
        }
        catch (e) { }
        try {
            const returnComparator = () => {
                return result.Certificate.Status;
            };
            if (returnComparator() === "FAILED") {
                return { state: utilWaiter.WaiterState.FAILURE, reason };
            }
        }
        catch (e) { }
    }
    catch (exception) {
        reason = exception;
        if (exception.name && exception.name == "ResourceNotFoundException") {
            return { state: utilWaiter.WaiterState.FAILURE, reason };
        }
    }
    return { state: utilWaiter.WaiterState.RETRY, reason };
};
const waitForCertificateValidated = async (params, input) => {
    const serviceDefaults = { minDelay: 60, maxDelay: 120 };
    return utilWaiter.createWaiter({ ...serviceDefaults, ...params }, input, checkState);
};
const waitUntilCertificateValidated = async (params, input) => {
    const serviceDefaults = { minDelay: 60, maxDelay: 120 };
    const result = await utilWaiter.createWaiter({ ...serviceDefaults, ...params }, input, checkState);
    return utilWaiter.checkExceptions(result);
};

Object.defineProperty(exports, "$Command", {
    enumerable: true,
    get: function () { return smithyClient.Command; }
});
Object.defineProperty(exports, "__Client", {
    enumerable: true,
    get: function () { return smithyClient.Client; }
});
exports.ACM = ACM;
exports.ACMClient = ACMClient;
exports.ACMServiceException = ACMServiceException;
exports.AccessDeniedException = AccessDeniedException;
exports.AddTagsToCertificateCommand = AddTagsToCertificateCommand;
exports.CertificateExport = CertificateExport;
exports.CertificateManagedBy = CertificateManagedBy;
exports.CertificateStatus = CertificateStatus;
exports.CertificateTransparencyLoggingPreference = CertificateTransparencyLoggingPreference;
exports.CertificateType = CertificateType;
exports.ConflictException = ConflictException;
exports.DeleteCertificateCommand = DeleteCertificateCommand;
exports.DescribeCertificateCommand = DescribeCertificateCommand;
exports.DomainStatus = DomainStatus;
exports.ExportCertificateCommand = ExportCertificateCommand;
exports.ExportCertificateRequestFilterSensitiveLog = ExportCertificateRequestFilterSensitiveLog;
exports.ExportCertificateResponseFilterSensitiveLog = ExportCertificateResponseFilterSensitiveLog;
exports.ExtendedKeyUsageName = ExtendedKeyUsageName;
exports.FailureReason = FailureReason;
exports.GetAccountConfigurationCommand = GetAccountConfigurationCommand;
exports.GetCertificateCommand = GetCertificateCommand;
exports.ImportCertificateCommand = ImportCertificateCommand;
exports.ImportCertificateRequestFilterSensitiveLog = ImportCertificateRequestFilterSensitiveLog;
exports.InvalidArgsException = InvalidArgsException;
exports.InvalidArnException = InvalidArnException;
exports.InvalidDomainValidationOptionsException = InvalidDomainValidationOptionsException;
exports.InvalidParameterException = InvalidParameterException;
exports.InvalidStateException = InvalidStateException;
exports.InvalidTagException = InvalidTagException;
exports.KeyAlgorithm = KeyAlgorithm;
exports.KeyUsageName = KeyUsageName;
exports.LimitExceededException = LimitExceededException;
exports.ListCertificatesCommand = ListCertificatesCommand;
exports.ListTagsForCertificateCommand = ListTagsForCertificateCommand;
exports.PutAccountConfigurationCommand = PutAccountConfigurationCommand;
exports.RecordType = RecordType;
exports.RemoveTagsFromCertificateCommand = RemoveTagsFromCertificateCommand;
exports.RenewCertificateCommand = RenewCertificateCommand;
exports.RenewalEligibility = RenewalEligibility;
exports.RenewalStatus = RenewalStatus;
exports.RequestCertificateCommand = RequestCertificateCommand;
exports.RequestInProgressException = RequestInProgressException;
exports.ResendValidationEmailCommand = ResendValidationEmailCommand;
exports.ResourceInUseException = ResourceInUseException;
exports.ResourceNotFoundException = ResourceNotFoundException;
exports.RevocationReason = RevocationReason;
exports.RevokeCertificateCommand = RevokeCertificateCommand;
exports.SortBy = SortBy;
exports.SortOrder = SortOrder;
exports.TagPolicyException = TagPolicyException;
exports.ThrottlingException = ThrottlingException;
exports.TooManyTagsException = TooManyTagsException;
exports.UpdateCertificateOptionsCommand = UpdateCertificateOptionsCommand;
exports.ValidationException = ValidationException;
exports.ValidationMethod = ValidationMethod;
exports.paginateListCertificates = paginateListCertificates;
exports.waitForCertificateValidated = waitForCertificateValidated;
exports.waitUntilCertificateValidated = waitUntilCertificateValidated;
